tests/test_kohya_loras_scaffold.py

```diff
@@ -0,0 +1,347 @@
+        scale = self.alpha / self.lora_dim
+        return self.multiplier * scale * self.lora_up(self.lora_down(x))
+
+
+class LoRAModuleContainer(torch.nn.Module):
+    def __init__(self, hooks, state_dict, multiplier):
+        super().__init__()
+        self.multiplier = multiplier
+
+        # Create LoRAModule from state_dict information
+        for key, value in state_dict.items():
+            if "lora_down" in key:
+                lora_name = key.split(".")[0]
+                lora_dim = value.size()[0]
+                lora_name_alpha = key.split(".")[0] + ".alpha"
+                alpha = None
+                if lora_name_alpha in state_dict:
+                    alpha = state_dict[lora_name_alpha].item()
+                if lora_name in hooks:
+                    hook = hooks[lora_name]
+                    lora_module = LoRAModule(hook.orig_module, lora_dim=lora_dim, alpha=alpha, multiplier=multiplier)
+                    self.register_module(lora_name, lora_module)
+
+        # Load whole LoRA weights
+        self.load_state_dict(state_dict, strict=False)
+
+        # Register LoRAModule to LoRAHook
+        for name, module in self.named_modules():
+            if module.__class__.__name__ == "LoRAModule":
+                if name in hooks:
+                    hook = hooks[name]
+                    hook.append_lora(module)
+
+    @property
+    def alpha(self):
+        return self.multiplier
+
+    @alpha.setter
+    def alpha(self, multiplier):
+        self.multiplier = multiplier
+        for name, module in self.named_modules():
+            if module.__class__.__name__ == "LoRAModule":
+                module.multiplier = multiplier
+
+    def remove_from_hooks(self, hooks):
+        for name, module in self.named_modules():
+            if module.__class__.__name__ == "LoRAModule":
+                hook = hooks[name]
+                hook.remove_lora(module)
+                del module
+
+
+class LoRAHook(torch.nn.Module):
+    """
+    replaces forward method of the original Linear,
+    instead of replacing the original Linear module.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.lora_modules = []
+
+    def install(self, orig_module):
+        assert not hasattr(self, "orig_module")
+        self.orig_module = orig_module
+        self.orig_forward = self.orig_module.forward
+        self.orig_module.forward = self.forward
+
+    def uninstall(self):
+        assert hasattr(self, "orig_module")
+        self.orig_module.forward = self.orig_forward
+        del self.orig_forward
+        del self.orig_module
+
+    def append_lora(self, lora_module):
+        self.lora_modules.append(lora_module)
+
+    def remove_lora(self, lora_module):
+        self.lora_modules.remove(lora_module)
+
+    def forward(self, x):
+        if len(self.lora_modules) == 0:
+            return self.orig_forward(x)
+        lora = torch.sum(torch.stack([lora(x) for lora in self.lora_modules]), dim=0)
```

Comment:
Beautiful! 

Seems like it should be able to handle multiple LoRAs, no?

Reply:
Yes, this hook version supports multiple LoRAs. Please refer to the comments in [this code](https://gist.github.com/takuma104/e38d683d72b1e448b8d9b3835f7cfa44) for usage. I would also like to see Diffusers support multiple LoRAs. I'm looking forward to future PRs for this.